docs/understand/kv-in-a-nutshell.md

```diff
@@ -0,0 +1,51 @@
+
+Under the hood, our KV implementation relies on a backing DB, which persists the data. Theoretically, it could be any type of database and out of the box, we already implemented drivers for [DynamoDB](https://en.wikipedia.org/wiki/Amazon_DynamoDB), for AWS users, and [PostgreSQL](https://en.wikipedia.org/wiki/PostgreSQL), using its relational nature to store a KV Store. More databases will be supported in the future, and lakeFS users and contributors can develop their own driver to use their own favorite database. For experimenting purposes, an in-memory KV store can be used, though it obviously lack the persistency aspect
+
+In order to store ref store objects (that is `Repositories`, `Branches`, `Commits`, `Tags`, and `Uncommitted Objects`), lakeFS implements another layer over the generic KV Store, which supports serialization and deserialization of these objects as [protobuf](https://en.wikipedia.org/wiki/Protocol_Buffers). As this layer relies on the generic interface of the KV Store layer, it is totally agnostic to whichever store implementation is in use, gaining our users the maximal flexibility
+
+For further reading, please refer to our [KV Design](https://github.com/treeverse/lakeFS/blob/master/design/accepted/metadata_kv/index.md)
+
+## Optimistic Locking with KV
+
+One important key difference between SQL databases and Key Value Store is the ability to lock resources. While this is a common practice with relational databases, Key Value stores lack this ability. Let us take a look at a common lakeFS flow, `Commit`, during which several ref store operations are performed:
+* All relevant (`Branch` correlated) uncommitted objects are collected and marked as committed 
+* A new `Commit` object is created
+* The relevant `Branch` is updated to point to the new commit
+
+The `Commit` flow includes multiple database accesses and modifications, and is very sensitive to concurrent executions: If 2 `Commit` flows run in parallel, we must guarantee correctness of the data. `lakeFS` with PostgreSQL simply locks the `Branch` for the entire `Commit` operation, preventing concurrent execution of such flows.
+Now, with KV Store replacing the SQL database, this easy solution is no longer available. Instead, we implemented an [Optimistic Locking](https://en.wikipedia.org/wiki/Optimistic_concurrency_control) algorithm, which leverages the KV Store `Compare-And-Set` (`CAS`) functionality to remember the `Branch` state at the beginning of the `Commit` flow, and updating the branch at the end, only if it remains unchanged, using `CAS`, with the former `Branch` state, used as a comparison criteria. If the sampled `Branch` state and the current state differ, it could only mean that another, later, `Commit` is in progress, causing the current `Commit` to fail, and give the later `Commit` a chance to complete.
```

Comment:
The second part of this paragraph could use some refactoring. I think referring to a 'later' commit might be a bit confusing. might want to talk about commit A and B instead?

Reply:
The commit A and B are used in the step-by-step example right after. The above is a general depction of a scenario that is later specified in greater detail